From f6093560b152a4d5ab69c36a71808c1bad9c2f1a Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:38:26 -0400
Subject: [ethernet] i40e: don't remove HMC that doesn't exist

Message-id: <1409658042-22689-146-git-send-email-sassmann@redhat.com>
Patchwork-id: 90340
O-Subject: [RHEL7.1 PATCH 145/281] i40e: don't remove HMC that doesn't exist
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Shannon Nelson <shannon.nelson@intel.com>

If a reset recovery failed (e.g. firmware is broken), the HMC space won't
get set up.  We don't need to try to delete it if it didn't get set up.
This stops some needless error messages when we already know we need to
just tear things down.

Change-ID: Iac600481765e20b136052b43a544e55d7870268b
Signed-off-by: Shannon Nelson <shannon.nelson@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 60442dea2e9314b472c772c76ed1b2ba5b961e02)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_main.c b/drivers/net/ethernet/intel/i40e/i40e_main.c
index 390a740..f5f2f50 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_main.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_main.c
@@ -5429,7 +5429,7 @@ static void i40e_fdir_teardown(struct i40e_pf *pf)
 static int i40e_prep_for_reset(struct i40e_pf *pf)
 {
  struct i40e_hw *hw = &pf->hw;
- i40e_status ret;
+ i40e_status ret = 0;
  u32 v;
 
  clear_bit(__I40E_RESET_INTR_RECEIVED, &pf->state);
@@ -5449,10 +5449,13 @@ static int i40e_prep_for_reset(struct i40e_pf *pf)
  i40e_shutdown_adminq(&pf->hw);
 
  /* call shutdown HMC */
- ret = i40e_shutdown_lan_hmc(hw);
- if (ret) {
-  dev_info(&pf->pdev->dev, "shutdown_lan_hmc failed: %d\n", ret);
-  clear_bit(__I40E_RESET_RECOVERY_PENDING, &pf->state);
+ if (hw->hmc.hmc_obj) {
+  ret = i40e_shutdown_lan_hmc(hw);
+  if (ret) {
+   dev_warn(&pf->pdev->dev,
+     "shutdown_lan_hmc failed: %d\n", ret);
+   clear_bit(__I40E_RESET_RECOVERY_PENDING, &pf->state);
+  }
  }
  return ret;
 }
@@ -8637,10 +8640,13 @@ static void i40e_remove(struct pci_dev *pdev)
  }
 
  /* shutdown and destroy the HMC */
- ret_code = i40e_shutdown_lan_hmc(&pf->hw);
- if (ret_code)
-  dev_warn(&pdev->dev,
-    "Failed to destroy the HMC resources: %d\n", ret_code);
+ if (pf->hw.hmc.hmc_obj) {
+  ret_code = i40e_shutdown_lan_hmc(&pf->hw);
+  if (ret_code)
+   dev_warn(&pdev->dev,
+     "Failed to destroy the HMC resources: %d\n",
+     ret_code);
+ }
 
  /* shutdown the adminq */
  ret_code = i40e_shutdown_adminq(&pf->hw);
-- 
1.7.1